darwin/future/core/types/common.py

```diff
@@ -73,18 +80,31 @@ class QueryString:
         Returns a string representation of the QueryString object, in the format "?key1=value1&key2=value2".
     """
 
-    value: Dict[str, str]
+    value: dict[str, list[str] | str]
 
-    def dict_check(self, value: UnknownType) -> Dict[str, str]:
-        assert isinstance(value, dict)
-        assert all(isinstance(k, str) and isinstance(v, str) for k, v in value.items())
-        return value
+    def dict_check(
+        self, value: Mapping[str, list[Stringable] | Stringable]
+    ) -> dict[str, list[str] | str]:
+        mapped: dict[str, list[str] | str] = {}
+        for k, v in value.items():
+            if isinstance(v, list):
+                mapped[k] = [str(x) for x in v]
+            else:
+                mapped[k] = str(v)
+        return mapped
 
-    def __init__(self, value: Dict[str, str]) -> None:
+    def __init__(self, value: Mapping[str, list[Stringable] | Stringable]) -> None:
         self.value = self.dict_check(value)
 
     def __str__(self) -> str:
-        return "?" + "&".join(f"{k}={v}" for k, v in self.value.items())
+        output: str = "?" if self.value else ""
+        for k, v in self.value.items():
+            if isinstance(v, list):
+                for x in v:
+                    output += f"{k}={x}&"
+            else:
+                output += f"{k}={v}&"
+        return output[:-1]  # remove trailing &
 
```

Comment:
This had to be changed and added to facilitate that a key can map to multiple values, eg dataset_id=0000&dataset_id=0001